File: backend-python/generation/generation_service.py
import asyncio
import uuid
import httpx
from fastapi import HTTPException
from config import config
from repositories.content_repository import content_repository

IMAGE_MODEL = 'black-forest-labs/FLUX.1-schnell-Free'
API_URL = 'https://api.together.xyz/v1/images/generations'


async def _generate_one(prompt: str) -> str:
    async with httpx.AsyncClient(timeout=60) as client:
        res = await client.post(API_URL, headers={
            'Authorization': f'Bearer {config.together_ai_api_key}',
            'Content-Type': 'application/json',
        }, json={
            'model': IMAGE_MODEL,
            'prompt': prompt,
            'width': 1024,
            'height': 1024,
            'steps': 4,
            'n': 1,
            'response_format': 'b64_json',
        })
    if not res.is_success:
        raise Exception(f'Together AI image generation failed: {res.text}')
    return res.json()['data'][0]['b64_json']


class GenerationService:
    async def generate_images(self, user_id: str, prompt: str, count: int = 4) -> list:
        safe_count = max(1, min(int(count or 1), 4))

        results = await asyncio.gather(
            *[_generate_one(prompt) for _ in range(safe_count)],
            return_exceptions=True,
        )

        images = []
        for res in results:
            if isinstance(res, Exception):
                print(f'[generation] One slot failed: {res}')
                continue
            record = content_repository.create({
                'id': str(uuid.uuid4()),
                'user_id': user_id,
                'prompt': prompt,
                'base64': res,
                'mime_type': 'image/jpeg',
            })
            images.append({
                'id': record['id'],
                'base64': record['base64'],
                'mimeType': record['mime_type'],
                'prompt': prompt,
            })

        if not images:
            raise HTTPException(status_code=502, detail='Image generation failed: no images returned from Together AI')
        return images

    async def regenerate_image(self, user_id: str, image_id: str, prompt: str) -> dict:
        existing = content_repository.find_by_id(image_id)
        if not existing or existing['user_id'] != user_id:
            raise HTTPException(status_code=404, detail='Image not found')
        base64 = await _generate_one(prompt)
        updated = content_repository.update(image_id, {'base64': base64, 'mime_type': 'image/jpeg', 'prompt': prompt})
        return {'id': updated['id'], 'base64': updated['base64'], 'mimeType': updated['mime_type'], 'prompt': prompt}

    def get_history(self, user_id: str) -> list:
        return [
            {
                'id': r['id'],
                'prompt': r['prompt'],
                'mimeType': r['mime_type'],
                'caption': r.get('caption'),
                'hashtags': r.get('hashtags'),
                'createdAt': r.get('created_at'),
                'updatedAt': r.get('updated_at'),
            }
            for r in content_repository.find_by_user_id(user_id)
        ]


generation_service = GenerationService()

File: backend/src/generation/generationService.js
import { GoogleGenerativeAI } from '@google/generative-ai';
import { v4 as uuidv4 } from 'uuid';
import config from '../config/index.js';
import { contentRepository } from '../repositories/contentRepository.js';

const genAI = new GoogleGenerativeAI(config.gemini.apiKey);

function getImageModel() {
  return genAI.getGenerativeModel({
    model: config.gemini.imageModel || 'gemini-2.5-flash-image',
  });
}

// Turns raw Gemini failure messages into a single user-facing error.
// Detects quota/rate-limit (429) so the cause is obvious instead of a generic 502.
function buildGenerationError(messages) {
  const joined = messages.join(' ');
  if (/\b429\b|quota|too many requests/i.test(joined)) {
    const err = new Error(
      'Gemini quota exceeded for the image model. The Nano Banana image models ' +
      'are not on the free tier — enable billing on your Google Cloud project, ' +
      'or set GEMINI_IMAGE_MODEL to a model your plan allows.'
    );
    err.status = 429;
    return err;
  }
  const err = new Error('Image generation failed: no images returned from Gemini');
  err.status = 502;
  return err;
}

export const generationService = {
  async generateImages({ userId, prompt, count = 4 }) {
    const safeCount = Math.min(Math.max(1, parseInt(count) || 1), 4);
    const model = getImageModel();

    const promises = Array.from({ length: safeCount }, () =>
      model.generateContent({
        contents: [{ role: 'user', parts: [{ text: prompt }] }],
        generationConfig: { responseModalities: ['IMAGE', 'TEXT'] },
      })
    );

    const responses = await Promise.allSettled(promises);
    const results = [];
    const failures = [];

    for (const res of responses) {
      if (res.status === 'rejected') {
        const msg = res.reason?.message || String(res.reason);
        console.error('[generation] Image generation failed for one slot:', msg);
        failures.push(msg);
        continue;
      }
      const candidate = res.value.response.candidates?.[0];
      if (!candidate) continue;
      const imagePart = candidate.content.parts.find(p => p.inlineData);
      if (!imagePart) continue;

      const record = contentRepository.create({
        id: uuidv4(),
        userId,
        prompt,
        base64: imagePart.inlineData.data,
        mimeType: imagePart.inlineData.mimeType || 'image/png',
      });
      results.push({
        id: record.id,
        base64: record.base64,
        mimeType: record.mimeType,
        prompt,
      });
    }

    if (results.length === 0) {
      throw buildGenerationError(failures);
    }
    return results;
  },

  async regenerateImage({ userId, imageId, prompt }) {
    const existing = contentRepository.findById(imageId);
    if (!existing || existing.userId !== userId) {
      const err = new Error('Image not found');
      err.status = 404;
      throw err;
    }
    const model = getImageModel();
    let response;
    try {
      response = await model.generateContent({
        contents: [{ role: 'user', parts: [{ text: prompt }] }],
        generationConfig: { responseModalities: ['IMAGE', 'TEXT'] },
      });
    } catch (err) {
      throw buildGenerationError([err?.message || String(err)]);
    }
    const imagePart = response.response.candidates?.[0]?.content.parts.find(p => p.inlineData);
    if (!imagePart) {
      const err = new Error('Regeneration failed: no image returned');
      err.status = 502;
      throw err;
    }
    const updated = contentRepository.update(imageId, {
      base64: imagePart.inlineData.data,
      mimeType: imagePart.inlineData.mimeType || 'image/png',
      prompt,
    });
    return { id: updated.id, base64: updated.base64, mimeType: updated.mimeType, prompt };
  },

  getHistory(userId) {
    return contentRepository.findByUserId(userId).map(r => ({
      id: r.id,
      prompt: r.prompt,
      mimeType: r.mimeType,
      caption: r.caption,
      hashtags: r.hashtags,
      createdAt: r.createdAt,
      updatedAt: r.updatedAt,
    }));
  },
};
